### PR TITLE
Kennric/client routes

### PR DESCRIPTION
**Pull request by Kennric**
fixes issue #20

## Changes in this PR.

- [X] Add spec for basic client CRUD endpoints
- [x] Implement CRUD for clients

## Testing this PR.

1. `docker-compose run dev bash` after making sure your app.env is correct
2. `rake spec`
3. See tests for Client crud run and pass

@osuosl/devs

### CODE REVIEW

**Comment by alxngyn** [comment]
+1, spec fails for me as expected
```
[root@923ba217bdbe code]# rake spec 

...

Finished in 0.14346 seconds (files took 0.77871 seconds to load)
20 examples, 6 failures

Failed examples:

rspec ./spec/routes/client_spec.rb:28 # The Clients endpoint displays a specific client by id
rspec ./spec/routes/client_spec.rb:39 # The Clients endpoint responds ok when asked for clients/new
rspec ./spec/routes/client_spec.rb:44 # The Clients endpoint responds ok when asked for the form to edit an existing client
rspec ./spec/routes/client_spec.rb:55 # The Clients endpoint allows us to create a new client, then redirects to the list
rspec ./spec/routes/client_spec.rb:69 # The Clients endpoint allows us to edit a client, then redirects to the list
rspec ./spec/routes/client_spec.rb:90 # The Clients endpoint allows us to edit a single client field then redirects to the list

/usr/local/bin/ruby -I/usr/local/lib/ruby/gems/2.3.0/gems/rspec-support-3.4.1/lib:/usr/local/lib/ruby/gems/2.3.0/gems/rspec-core-3.4.4/lib 
/usr/local/lib/ruby/gems/2.3.0/gems/rspec-core-3.4.4/exe/rspec spec/models/client_spec.rb spec/routes/app_spec.rb spec/routes/client_spec.rb --format documentation failed

```

**Comment by ElijahCaine** [comment]
+1 I also get failures.

**Comment by Kennric** [comment]
Code now implements basic crud. This PR does not provide html forms for this, just the underlying functionality, we will want capybara or similar integration testing for that.

File: spec/routes/client_spec.rb [hunk 1]
```diff
@@ -1,23 +1,106 @@
...
+    expect(last_response.status).to eq(200)
+  end
+
+  it 'responds 404 when asked for the form to edit an absent client' do
+    get '/clients/9884093/edit'
+    expect(last_response.status).to eq(404)
+  end
+
+  it 'allows us to create a new client, then redirects to the list' do
+    post '/clients', name: 'I\'m new!'
```
Comment (ElijahCaine) [inline spec/routes/client_spec.rb]: I was under the impression that any string with `'` was to be double quoted; `"I'm new!"` for instance. Escaping characters is pretty ugly imo.

**Comment by ElijahCaine** [comment]
Tests pass (`20 examples, 0 failures` and `3 examples, 0 failures` :smile: )

The code looks good. Tests are very expressive and pretty simple. I think they cover all the major bases for CRUD and serve as a good template for future CRUD we'll be working on.

One point I'd like to consider is storing all endpoints in `app.rb` or storing each endpoint in it's own file (in an `endpoints/` directory or something) and extending the `Iam` class somehow. I'm not sure the best way to go about this, I just don't want `app.rb` to get too unwieldy. If it's not a problem I'm 100% okay with writing everything in `app.rb`, especially if that's best practice.

**Comment by Kennric** [comment]
@ElijahCaine many sinatra developers split out all the routes (or 'controllers' depending on the term you like) into an app/ directory, with models/ controllers/ views/ etc under it. That seemed a little overkill to me, but I am not opposed to putting each object's endpoints into their own files, I believe that is fairly easy. When we get into complicated code for the reporting and other things, we'll definitely want to split things into other locations anyway. Let's keep it in app for now, and discuss ways to split things out as a separate issue.

File: spec/routes/client_spec.rb (hunk 1 above)
Comment (Kennric) [inline spec/routes/client_spec.rb]: Fixed - that was from an overzealous round of rubocop fixes.

Comment (matthewrsj) [inline spec/routes/client_spec.rb]: This is still `\'`

File: spec/routes/client_spec.rb [hunk 2]
```diff
@@ -1,23 +1,106 @@
...
+
+    client = Client[name: 'I\'m new!']
```
Comment (matthewrsj) [inline spec/routes/client_spec.rb]: This is still `\'`

**Comment by matthewrsj** [comment]
Aside from the nitpick on the `\'`s above, +1.

**Comment by ElijahCaine** [comment]
+1. I went ahead and made that nit-pick change and re-reviewed things. Any final thoughts @matthewrsj ?

**Comment by matthewrsj** [comment]
+1

**Pull request merged**